strings/base_iterator.h

```diff
@@ -0,0 +1,183 @@
+
+namespace winrt::impl
+{
+    template <typename T>
+    struct fast_iterator
+    {
+        using iterator_concept = std::random_access_iterator_tag;
+        using iterator_category = std::input_iterator_tag;
+        using value_type = decltype(std::declval<T>().GetAt(0));
+        using difference_type = ptrdiff_t;
+        using reference = value_type;
+
+        fast_iterator() noexcept : m_collection(nullptr), m_index(0) {}
+
+        fast_iterator(T const& collection, uint32_t const index) noexcept :
+            m_collection(&collection),
+            m_index(index)
+        {}
+
+        fast_iterator& operator++() noexcept
+        {
+            ++m_index;
+            return *this;
+        }
+
+        fast_iterator operator++(int) noexcept
+        {
+            auto previous = *this;
+            ++m_index;
+            return previous;
+        }
+
+        fast_iterator& operator--() noexcept
+        {
+            --m_index;
+            return *this;
+        }
+
+        fast_iterator operator--(int) noexcept
+        {
+            auto previous = *this;
+            --m_index;
+            return previous;
+        }
+
+        fast_iterator& operator+=(difference_type n) noexcept
+        {
+            m_index += static_cast<uint32_t>(n);
+            return *this;
+        }
+
+        fast_iterator operator+(difference_type n) const noexcept
+        {
+            return fast_iterator(*this) += n;
+        }
+
+        fast_iterator& operator-=(difference_type n) noexcept
+        {
+            return *this += -n;
+        }
+
+        fast_iterator operator-(difference_type n) const noexcept
+        {
+            return *this + -n;
+        }
+
+        difference_type operator-(fast_iterator const& other) const noexcept
+        {
+            return static_cast<difference_type>(m_index) - static_cast<difference_type>(other.m_index);
+        }
+
+        reference operator*() const
+        {
+            return m_collection->GetAt(m_index);
+        }
+
+        reference operator[](difference_type n) const
+        {
+            return m_collection->GetAt(m_index + static_cast<uint32_t>(n));
+        }
+
+        bool operator==(fast_iterator const& other) const noexcept
+        {
+            WINRT_ASSERT(m_collection == other.m_collection);
```

Comment:
Should this assert be applied to `operator-` and friends?

Reply:
I see no reason not to, applied.